### PR TITLE
Issue: Incorrect Data Center for the "Upload Image" dialog

### PR DESCRIPTION
**Pull request by sermakov-orion**
Signed-off-by: Stepan Ermakov <sermakov@orionsoft.ru>

## Changes introduced with this PR
This PR fixes the following issue: incorrect Data Center is displayed when uploading a disk image from "Storage Domain Disks" UI.
**Steps to reproduce:**
    1. Create a Storage Domain at a non-Default Data Center.
    2. Open the Storage -> Domains -> [The created Storage Domain] -> Disks page.
    3. Select the Upload -> Start menu item. The "Upload image" dialog is displayed.

**Expected behavior:**
    The Data Center field is disabled and displays the name of the non-Default Data Center.
**Current behavior:**
    The Data Center field is disabled and displays the name of the Default Data Center.

Fixed in the following way: when "Upload image" dialog is displayed then query the Storage Domain details with information about its Data Center and select the Storage Domain's Data Center in the "Data Center" field

## Are you the owner of the code you are sending in, or do you have permission of the owner?
y

### CODE REVIEW

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 1]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
 
                     getDataCenter().setIsChangeable(isChangeable);
                     if (!isChangeable) {
-                        // Set the selected item on the storage domains list.
-                        AsyncDataProvider.getInstance().getStorageDomainById(
-                                new AsyncQuery<>(storageDomain -> {
-                                    getStorageDomain().setItems(Collections.singletonList(storageDomain));
-                                }), limitToStorageDomainId);
```
Comment (sermakov-orion) [inline frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java:176]: `AsyncDataProvider.getInstance().getStorageDomainById ` queries Storage Domain details by using the `GetStorageDomainById` query which in turn uses the `getstorage_domains_by_id` function which uses `storage_domains_without_storage_pools` view. And the `storage_domains_without_storage_pools` view does not return the `storagePoolId` (Data Center id) property for the Storage Domain

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 2]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
...
+                        // Set the selected item on the storage domains and data centers lists.
+                        Frontend.getInstance().runQuery(
+                                QueryType.GetStorageDomainListById,
+                                new IdQueryParameters(limitToStorageDomainId),
+                                new AsyncQuery<QueryReturnValue>(returnValue -> {
```
Comment (sermakov-orion) [inline frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java:181]: So, to get the Data Center Id we have to use another Query (`GetStorageDomainListById`)

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 3]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
...
+                            ArrayList<StorageDomain> domains = returnValue.getReturnValue();
+                            if (!domains.isEmpty()) {
+                                StorageDomain storageDomain = domains.get(0);
```
Comment (sermakov-orion) [inline frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java:184]: which in general may return a list of Storage Domains - it looks that this is a pretty old behavior that is currently changed, so we can consider only one Storage Domain in the result.

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 4]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
...
+                                // Set the selected item on the storage domains list.
+                                getStorageDomain().setItems(Collections.singletonList(storageDomain));
+                                Guid dcId = storageDomain.getStoragePoolId();
```
Comment (sermakov-orion) [inline frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java:187]: And the result also contains the `storagePoolId` field, so we can use it to display proper Data Center in the appropriate field of the "Upload Image" Dialog.

**Comment by sermakov-orion** [comment]
Requesting Reviews.

**Review by barpavel — COMMENTED** [review]
I admit I didn't invest a lot of time in this code-review and the memories of working on `ovirt-engine` are not that fresh already :)
Just a few things to think about, please disregard if they are not relevant.

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 5]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
...
+                                if (dcId != null) {
```
[the remark this answers is not in the document]
Comment (sermakov-orion) [reply frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java]: @barpavel 
Changed to Guid.isNullOrEmpty()

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 6]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
...
+                                    // Set selected data center if the list of data centers is populated.
+                                    selectDataCenter(dcId);
+                                    // Or listen to the data centers list change to update selected data center.
+                                    getDataCenter().getItemsChangedEvent()
+                                            .addListener((ev, sender, args) -> selectDataCenter(dcId));
```
[the remark this answers is not in the document]
Comment (sermakov-orion) [reply frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java]: @barpavel 
In general, the initial version works with no issues. But I agree - in some cases, when "the above selectDataCenter(dcId); (line 190) succeeded" then there is no need to configure the listener.
So, I refactored the code and adjusted the comments. Please take a look.

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java [hunk 7]
```diff
@@ -169,11 +174,26 @@ public void initialize() {
...
+                        // Set the selected item on the storage domains and data centers lists.
+                        Frontend.getInstance().runQuery(
+                                QueryType.GetStorageDomainListById,
+                                new IdQueryParameters(limitToStorageDomainId),
+                                new AsyncQuery<QueryReturnValue>(returnValue -> {
+                            ArrayList<StorageDomain> domains = returnValue.getReturnValue();
+                            if (!domains.isEmpty()) {
+                                StorageDomain storageDomain = domains.get(0);
+                                // Set the selected item on the storage domains list.
+                                getStorageDomain().setItems(Collections.singletonList(storageDomain));
```
[the remark this answers is not in the document]
Comment (sermakov-orion) [reply frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java]: @barpavel 
Collections.singletonList works fine in this place. Also I found many cases of use of Collections.singletonList in the **uicommonweb** module. But at the same time no any case with the ListModel.setItems. So, finally, I decided to agree with your concern. And changed this to the ArrayList.
Thanks,

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (barpavel) [reply frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java]: Can be resolved then :)

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (barpavel) [reply frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java]: This comment can be resolved then, thank you!

**Comment by barpavel** [comment]
LGTM.

File: frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sermakov-orion) [reply frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java]: @barpavel 
Comments - fixed
stream-filter-findfirst - fixed
Closing the conversation... Thanks.

**Review by sandrobonazzola — APPROVED** [review]
Approving based on @barpavel review

**Pull request merged**